src/runtime/HalideBuffer.h

```diff
@@ -2010,6 +2010,19 @@ class Buffer {
     template<int N>
     HALIDE_NEVER_INLINE static bool for_each_value_prep(for_each_value_task_dim<N> *t,
                                                         const halide_buffer_t **buffers) {
+        // Check the buffers all have clean host allocations
+        for (int i = 0; i < N; i++) {
+            if (buffers[i]->device) {
+                assert(!buffers[i]->device_dirty() &&
```

Comment:
I think if you flip these two asserts you can eliminate the suggestion to call "allocate" in the device_dirty case.

Reply:
Good point. Done